int64 id
float64 p
int64 dir